tiledb/sm/array/array_operations.cc

```diff
@@ -0,0 +1,116 @@
+/**
+ * @file   array_operations.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements I/O operations which support the Array class.
+ *
+ * The data- and I/O-oriented operations of class `Array` have been
+ * intentionally separated to support two _standalone_ object libraries.
+ * This design eliminates cyclic dependencies in the build, as each module has
+ * its own dependency chain.
+ * Note that the `array_operations` object library will not be implemented
+ * until _after_ the `Query` object libraries on which it relies.
+ */
+
+#include "tiledb/sm/array/array_operations.h"
+#include "tiledb/common/stdx_string.h"
+#include "tiledb/sm/array/array.h"
+#include "tiledb/sm/array/array_directory.h"
+#include "tiledb/sm/crypto/encryption_key.h"
+#include "tiledb/sm/misc/parallel_functions.h"
+#include "tiledb/sm/query/deletes_and_updates/serialization.h"
+#include "tiledb/sm/tile/generic_tile_io.h"
+
+namespace tiledb::sm {
+
+class ArrayOperationsException : public StatusException {
+ public:
+  explicit ArrayOperationsException(const std::string& message)
+      : StatusException("ArrayOperations", message) {
+  }
+};
+
+/* ********************************* */
+/*                API                */
+/* ********************************* */
+
+tuple<std::vector<QueryCondition>, std::vector<std::vector<UpdateValue>>>
+load_delete_and_update_conditions(
+    ContextResources& resources, const shared_ptr<OpenedArray>& opened_array) {
```

Comment:
```suggestion
    ContextResources& resources, OpenedArray& opened_array) {
```

We're not going to keep any persistent reference to `opened_array`, so there's no reason to reference the `shared_ptr`. Referencing the actual object is better.

I suspect it could be a constant reference, but I didn't trace it out at all.